nearcore/src/dyn_config.rs

```diff
@@ -0,0 +1,73 @@
+use crate::config::Config;
+use near_chain_configs::UpdateableClientConfig;
+use near_dyn_configs::{DynConfigsError, UpdateableConfigs};
+use near_o11y::log_config::LogConfig;
+use serde::Deserialize;
+use std::path::{Path, PathBuf};
+
+const LOG_CONFIG_FILENAME: &str = "log_config.json";
+
+/// This function gets called at the startup and each time a config needs to be reloaded.
+pub fn read_updateable_configs(home_dir: &Path) -> Result<UpdateableConfigs, DynConfigsError> {
+    let mut errs = vec![];
+    let log_config = match read_log_config(home_dir) {
+        Ok(config) => config,
+        Err(err) => {
+            errs.push(err);
+            None
+        }
+    };
+    let updateable_client_config =
```

Comment:
I think it is better UX if the updateable configs are read through a separate file instead of the original config file. It has a few benefits:
- Users won't be able to modify a config field that is not actually upgradeable. Right now, if user does that, the code just silently drops the update, which could be a source of confusion to the user. On top of that, the node can be in trouble when they restart. 
- Users can't accidentally change the config file to a bad state, so they won't be in trouble when they restart the node.

Reply:
IIUC, the suggestion is to have `config.json` which is assumed to be in a good state; and to duplicate the updateable fields into a separate file, e.g. `dyn_config.json`.

If in the future we decide to make `max_num_peers` updateable at runtime, will that mean that the same field will be configurable in `config.json` as `network.max_num_peers` and in `dyn_config.json` as `max_num_peers`?

I would find that UX confusing, because reading `config.json` is no longer enough to understand the node's configuration. There is always a chance that some field is overridden in `dyn_config.json`.
On the other hand, if I want to change `max_num_peers`, why shouldn't I change it in both `config.json` and `dyn_config.json`?
Which means that to read the configuration I need to read two config files; and to update the configuration I need to update two config files.

Reply:
I think my main concern with let user update config.json directly is that in that case, the values in config.json may not actually reflect what the config the node is using. But I also see your point about two sources of truth. 

What about we let the user update the dyn_config.json but the code will automatically update `config.json` to reflect the updated value after dyn_config.json is applied. This way, the user only needs to rely on `config.json` for the source of truth. And they don't have to worry about messing up the `config.json` when the node is running.

Reply:
To make it even safer, the program can also save a copy of the old config value, so the user has something to fall back to in case they make a mistake and want to revert to the previous config.

Reply:
Just an idea about this concern.  Could the following rules be an easier solution?   

- If the dyn_config is valid and applied.  Could we save a config snapshot `config.{timesampts}.json`? 
- If the snapshot can not store, neard will panic in case of any config changes without a clear snapshot.
- Also, there could be a field `snapshot_config_change: false` (default is true) in the original `config.json`, and this value is not allowed changed in runtime.  Just in case that people do not want these snapshots.

If these look good from your end, I am happy to implement this. :smile:

Reply:
@mzhangmzz 

> the code will automatically update config.json

I find it very confusing when a program modifies its own configs.

> the user has something to fall back to in case they make a mistake

Unclear how useful this is. I rely on the (persistent) undo history. Some users make and manage backups of the configs.

Reply:
@yanganto 

> `allow_config_change`

Seems redundant when a `SIGHUP` is required to trigger a config reload. If the user doesn't send a signal, nothing will happen.

> save a config snapshot

I believe many users would like to manage config versions themselves.

Reply:
I also like the idea of a single 'config.json' file.

For the config{timestamp}.json - the part I don't like, is that it pollutes the directory a little - what we could do instead, is to have the /config HTTP endpoint (or /debug/config) that would always report the current config.  (this will allow people to do diffs in an easy way if needed)

Reply:
`/config` is already work in progress: https://pagodaplatform.atlassian.net/browse/ND-273